File: oa2-examples/python/fetch_all_records.py
#!/usr/bin/env python3
"""
Print every OpenAlias v2 record published for an alias.

Fetches and prints the `_openalias-payment` and `_openalias-metadata` records,
using the DNSSEC-authenticated resolution in openalias.py.

Requires: dnspython with cryptography  ->  pip install "dnspython[dnssec]"

Usage:
    python3 python-fetch-all-records.py donate@openalias.org
"""

import sys

import openalias


def main(alias: str) -> int:
    try:
        fqdn = openalias.normalize(alias)
    except openalias.NotAnAlias as exc:
        print(exc)
        return 1

    print(f"{alias}  ->  {fqdn}\n")

    exit_code = 0
    for prefix in (openalias.PAYMENT_PREFIX, openalias.METADATA_PREFIX):
        name = f"{prefix}.{fqdn}"
        try:
            found = openalias.records(alias, prefix)
        except openalias.InsecureLookup as exc:
            print(f"{name}: REFUSED — {exc}\n")
            exit_code = 1
            continue
        except openalias.LookupFailed as exc:
            print(f"{name}: lookup failed — {exc}\n")
            exit_code = 1
            continue

        if not found:
            print(f"{name}: no OpenAlias v2 records\n")
            continue

        print(f"{name}: {len(found)} record(s)")
        for fields in found:
            for key, value in fields.items():
                print(f"    {key} = {value}")
            print()

    return exit_code


if __name__ == "__main__":
    if len(sys.argv) != 2:
        sys.exit(f"usage: {sys.argv[0]} <alias>   e.g. donate@openalias.org")
    sys.exit(main(sys.argv[1]))

File: oa2-examples/dart/fetch_all_records.dart
// Print every OpenAlias v2 record published for an alias.
//
// A Dart port of the Python fetch-all example. Requires the native helper to be
// built first: `cd native && cargo build --release`.
//
// Usage:
//   dart run fetch_all_records.dart donate@openalias.org

import 'dart:io';

import 'openalias.dart' as oa;

void main(List<String> args) {
  if (args.length != 1) {
    stderr.writeln('usage: dart run fetch_all_records.dart <alias>   e.g. donate@openalias.org');
    exit(1);
  }
  final alias = args[0];

  final String fqdn;
  try {
    fqdn = oa.normalize(alias);
  } on oa.NotAnAlias catch (e) {
    print(e);
    exit(1);
  }

  print('$alias  ->  $fqdn\n');

  for (final prefix in [oa.paymentPrefix, oa.metadataPrefix]) {
    final name = '$prefix.$fqdn';
    try {
      final found = oa.records(alias, prefix);
      if (found.isEmpty) {
        print('$name: no OpenAlias v2 records\n');
        continue;
      }
      print('$name: ${found.length} record(s)');
      for (final fields in found) {
        fields.forEach((key, value) => print('    $key = $value'));
        print('');
      }
    } on oa.InsecureLookup catch (e) {
      print('$name: REFUSED — $e\n');
    } on oa.LookupFailed catch (e) {
      print('$name: lookup failed — $e\n');
    }
  }
}
